src/combinatorics/simple_graph/adj_matrix.lean

```diff
@@ -15,89 +16,247 @@ properties to computational properties of the matrix.
 
 ## Main definitions
 
-* `adj_matrix` is the adjacency matrix of a `simple_graph` with coefficients in a given semiring.
+* `matrix.is_adj_matrix`: `A : matrix V V α` is qualified as an "adjacency matrix" if
+  (1) every entry of `A` is `0` or `1`,
+  (2) `A` is symmetric,
+  (3) every diagonal entry of `A` is `0`.
+
+* `matrix.is_adj_matrix.to_graph`: for `A : matrix V V α` and `h : A.is_adj_matrix`,
+  `h.to_graph` is the simple graph induced by `A`.
+
+* `matrix.compl`: for `A : matrix V V α`, `A.compl` is supposed to be
+  the adjacency matrix of the complement graph of the graph induced by `A`.
+
+* `simple_graph.adj_matrix`: the adjacency matrix of a `simple_graph`.
 
 -/
 
 open_locale big_operators matrix
 open finset matrix simple_graph
 
-universes u v
-variables {α : Type u}
-variables (R : Type v) [semiring R]
+variables {V α β : Type*}
+
+namespace matrix
+
+/-- `A : matrix V V α` is qualified as an "adjacency matrix" if
+    (1) every entry of `A` is `0` or `1`,
+    (2) `A` is symmetric,
+    (3) every diagonal entry of `A` is `0`. -/
+structure is_adj_matrix [has_zero α] [has_one α] (A : matrix V V α) : Prop :=
+(zero_or_one : ∀ i j, (A i j) = 0 ∨ (A i j) = 1 . obviously)
+(symm : A.is_symm . obviously)
+(loopless : ∀ i, A i i = 0 . obviously)
+
+namespace is_adj_matrix
+
+variables {A : matrix V V α}
+
+lemma apply_diag [has_zero α] [has_one α] (h : is_adj_matrix A) (i : V) :
+  A i i = 0 :=
+by simp [h.loopless]
+
+@[simp]
+lemma apply_diag_ne [mul_zero_one_class α] [nontrivial α] (h : is_adj_matrix A) (i : V) :
+  ¬ A i i = 1 :=
+by simp [h.apply_diag i]
+
+@[simp]
+lemma apply_ne_one_iff [mul_zero_one_class α] [nontrivial α] (h : is_adj_matrix A) (i j : V) :
+  ¬ A i j = 1 ↔ A i j = 0 :=
+begin
+  replace h := h.zero_or_one i j,
+  split,
+  { tauto },
+  { rintros g, simp [g] }
+end
+
+@[simp]
+lemma apply_ne_zero_iff [mul_zero_one_class α] [nontrivial α] (h : is_adj_matrix A) (i j : V) :
+  ¬ A i j = 0 ↔ A i j = 1 :=
+by { rw [←apply_ne_one_iff h], simp }
+
+/-- For `A : matrix V V α` and `h : is_adj_matrix A`,
+    `h.to_graph` is the simple graph whose adjacency matrix is `A`. -/
+def to_graph [mul_zero_one_class α] [nontrivial α] (h : is_adj_matrix A) :
```

Comment:
If you add `@[simps]`, it will automatically create a `simp` lemma named `matrix.is_adj_matrix.to_graph_adj` that says `h.to_graph.adj i j = (A i j = 1)`